arangod/Aql/Ast.cpp

```diff
@@ -327,6 +327,57 @@ LogicalDataSource::Category injectDataSourceInQuery(
   return dataSource->category();
 }
 
+LogicalDataSource::Category addDataSource(
+    Ast& ast, CollectionNameResolver const& resolver, std::string name) {
+  std::string_view nameRef(name);
+
+  auto category = injectDataSourceInQuery(ast, resolver, AccessMode::Type::READ,
+                                          false, nameRef);
+
+  if (category == LogicalDataSource::Category::kCollection) {
+    if (ServerState::instance()->isCoordinator()) {
+      auto& ci = ast.query()
+                     .vocbase()
+                     .server()
+                     .getFeature<ClusterFeature>()
+                     .clusterInfo();
+      auto c = ci.getCollectionNT(ast.query().vocbase().name(), name);
+      if (c != nullptr) {
+        auto const& names = c->realNames();
+
+        for (auto const& n : names) {
+          std::string_view shardsNameRef(n);
+
+          LogicalDataSource::Category shardsCategory = injectDataSourceInQuery(
+              ast, resolver, AccessMode::Type::READ, false, shardsNameRef);
+
+          TRI_ASSERT(shardsCategory ==
+                     LogicalDataSource::Category::kCollection);
+        }
+      }  // else { TODO Should we really not react? }
+    }
+  }
+
+  return category;
+}
+
+std::optional<std::string> edgeCollectionNodeGetName(
+    CollectionNameResolver const& resolver, AstNode const* edgeCollection) {
+  if (edgeCollection->isStringValue()) {
+    return edgeCollection->getString();
+  } else if (edgeCollection->type == NODE_TYPE_DIRECTION) {
+    TRI_ASSERT(edgeCollection->numMembers() == 2);
+    auto eCSub = edgeCollection->getMember(1);
+
+    if (eCSub->isStringValue()) {
+      return eCSub->getString();
+    }
+  }
+
+  // TODO: find whether this can ever happen
+  return std::nullopt;
```

Comment:
Should we add a `TRI_ASSERT(false)` here? That way we should be able to find if this can actually happen. ;)